### PR TITLE
Fixup call `ExecutionSpace::concurrency()` from an actual execution space

### PR DESCRIPTION
**Pull request by dalg24**
Partially addressing #5655 by ensuring we call `ExecutionSpace::concurrency()` from an actual execution space in code that is not backend-specific.

Also adding the detection of the member function to static checks for all backends.  Note that I elected to only check that `concurrency()` is callable and that, even though that is something that we would be able to check later after we update all backends, I don't actually plan on enforcing that backends do not declare the member function to be static.

### CODE REVIEW

File: core/perf_test/PerfTest_CustomReduction.cpp [hunk 1]
```diff
@@ -58,8 +58,8 @@ void custom_reduction_test(int N, int R, int num_trials) {
   Scalar max;
 
   int team_size = 32;
-  if (team_size > Kokkos::DefaultExecutionSpace::concurrency())
-    team_size = Kokkos::DefaultExecutionSpace::concurrency();
+  if (team_size > Kokkos::DefaultExecutionSpace().concurrency())
+    team_size = Kokkos::DefaultExecutionSpace().concurrency();
```
Comment (PhilMiller) [inline core/perf_test/PerfTest_CustomReduction.cpp:62]: This raises a question: how confident are we that we can generally promise that constructing a default instance of an execution space is and will remain 'cheap'? It's something that happens in a lot of places, but it's never been clear to me if that's actually a commitment Kokkos makes.

Comment (masterleinad) [reply core/perf_test/PerfTest_CustomReduction.cpp:62]: I mean that's the reason why we are using a static (`Impl`) execution space object for a default-constructed execution space instance.

**Comment by dalg24** [comment]
Failure is unrelated (`KokkosCore_UnitTest_CudaTimingBased` in the NVHPC build)
```
7: [ RUN      ] cuda.debug_pin_um_to_host
7: Time CudaSpace: 0.006675 CudaUVMSpace_1: 0.011526 CudaUVMSpace_2: 0.006572 CudaPinnedHostSpace: 0.269566 CudaUVMSpace_Pinned: 0.019738
7: /var/jenkins/workspace/Kokkos/core/unit_test/cuda/TestCuda_DebugPinUVMSpace.cpp:128: Failure
7: Value of: passed
7:   Actual: false
7: Expected: true

**Comment by dalg24** [comment]
I forgot to mention this PR introduces conflicts with #5598
They are no big deal but I wanted to acknowledge I was aware of these.  @tcclevenger

**Review by ldh4 — APPROVED** [review]
Looks fine to me.

**Review by crtrott — APPROVED** [review]

**Pull request merged**